src/main/java/com/team766/controllers/CSMSMPID.java

```diff
@@ -0,0 +1,214 @@
+			}catch (IllegalArgumentException ill){
+				throw new Exception("Error instantiating the CLE PID controller: " + ill);
+			}
+			
+			
+	}
+	//manually changing the state
+	public void updateState(PIDSTATE state){
+		theState = state;
+	}
+	//changing all PID values at once
+	public void setPIDF(double p, double i, double d, double ff){
+		pid1.setP(p);
+		pid1.setI(i);
+		pid1.setD(d);
+		pid1.setFF(ff);
+	}
+	//changing the P value
+	public void setP(double p){
+		pid1.setP(p);
+	}
+	//changing the I value
+	public void setI(double i){
+		pid1.setI(i);
+	}
+	//changing the D value
+	public void setD(double d){
+		pid1.setD(d);
+	}
+	//changing the FF value
+	public void setFf(double ff){
+		pid1.setFF(ff);
+	}
+	
+	//setting the antigravity constants2
+	public void setAntigravConstant(double k){
+		antiGravK = k;
+	}
+
+	private void antigrav(){
+		mc1.set(antiGravK * Math.sin(mc1.getSensorPosition()));
+	}
+
+	//adding a built in closed loop error (not tested yet)
+	public void setSmartMotionAllowedClosedLoopError(double error){
+		pid1.setSmartMotionAllowedClosedLoopError(error, 0);
+	}
+	//changing the deadzone
+	public void setDeadzone(double dz){
+		dz1 = dz;
+	}
+	//changing the output range of the speed of the motors
+	public void setOutputRange(double min, double max){
+		maxspeed1 = max;
+		minspeed1 = min;
+		pid1.setOutputRange(min, max);
+	}
+	//changing the neutral mode of the motor (brake/coast)
+	public void setMotorMode(NeutralMode mode){
+		mc1.setNeutralMode(mode);
+	}
+	//setting the maximum and minimul locations that the motor can go to
+	public void setMinMaxLocation(double min, double max){
+		maxpos1 = max;
+		minpos1 = min;
+	}
+	//setting the maximum velocity of the motor
+	public void setMaxVel(double max){
+		maxvel1 = max;
+		pid1.setSmartMotionMaxVelocity(max, 0);
+		pid1.setSmartMotionMinOutputVelocity(0, 0);
+	}
+	//setting the maximum acceleration of the motor
+	public void setMaxAccel(double max){
+		maxaccel1 = max;
+		pid1.setSmartMotionMaxAccel(max, 0);
+	}
+
+	//go to a position using the thing that wasn't tested yet and almost broke the robot...
+	public void setCLEPosition(double position){
+		if(position > maxpos1){
+			position = maxpos1;
+		} else if(position < minpos1){
+			position = minpos1;
+		}
+
+		pid1.setReference(position, ControlType.kSmartMotion);
+	}
+	//1st step to go to a position using the normal PID, setting what you want the position to be
+	public void setPosition(double position){
+		if(position > maxpos1){
+			position = maxpos1;
+		} else if(position < minpos1){
+			position = minpos1;
+		}
+
+		currentPos = position;
+	}
+
+	public void stop(){
+		//Failsafe
+		currentPos = mc1.getSensorPosition();
+		theState = PIDSTATE.OFF;
+	}
+
+	//run loop that actually runs the PID using normal dz
+	//You need to call this function repedatly in OI
+	public void run(boolean enabled){
+		if(enabled){
+			switch(theState){
+				case OFF:
+					break;
+				case ANTIGRAV:
+					antigrav();
```

Comment:
the antigrav procedure applies force to counteract gravity, but a potentially significant in this control scheme is that it does not resist dynamic loads on the mechanism: such is it getting pushed by an external force, or the effects of inertia when the robot accelerates or decelerates.

One way to address this would be to transition back to the `PID` state if the mechanism moves out of the deadzone around the setpoint

Reply:
At least for this year's robot, it goes to antigrav and remains there, until the next movement is desired.
I agree it should switch back to pid, but needs testing.

Reply:
I've added this, let me know how it looks